### PR TITLE
DAC6-40 Create postcode lookup address page

### PR DESCRIPTION
**Pull request by helensoniega**

### CODE REVIEW

File: app/connectors/AddressLookupConnector.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    http.GET[HttpResponse](
+      url = addressLookupUrl,
+      queryParams = urlParams,
+      headers = Seq("X-Hmrc-Origin" -> "DAC6")
+    )(implicitly, implicitly, implicitly) flatMap {
+      case response if response.status equals OK =>
+        Future.successful(response.json.as[Seq[AddressLookup]].filterNot(address =>
+          address.addressLine1.isEmpty && address.addressLine2.isEmpty && address.addressLine3.isEmpty &&
+            address.addressLine4.isEmpty)
```
Comment (daniel-manning) [inline app/connectors/AddressLookupConnector.scala:47]: seems really restrictive, I don't know how we collect the data but not everyone puts in a 3rd or 4th line when sending address details

File: app/connectors/AddressLookupConnector.scala [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    val addressLookupUrl: String = s"${config.addressLookUpUrl}/v2/uk/addresses"
+    val urlParams = Seq ("postcode" -> postCode)
+
+    implicit val reads: Reads[Seq[AddressLookup]] = AddressLookup.addressesLookupReads
+
+    http.GET[HttpResponse](
+      url = addressLookupUrl,
+      queryParams = urlParams,
+      headers = Seq("X-Hmrc-Origin" -> "DAC6")
+    )(implicitly, implicitly, implicitly) flatMap {
```
Comment (daniel-manning) [inline app/connectors/AddressLookupConnector.scala:43]: if you're passing everything with the implicitly method it means the compiler should be able to resolve them all too

File: app/connectors/AddressLookupConnector.scala [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      case response if response.status equals OK =>
+        Future.successful(response.json.as[Seq[AddressLookup]].filterNot(address =>
```
Comment (daniel-manning) [inline app/connectors/AddressLookupConnector.scala:45]: they've just changed how httpreads work, when we shift to the new play bootstrap we can clean this up

File: app/controllers/IndividualUKPostcodeController.scala [hunk 4]
```diff
@@ -48,18 +50,21 @@ class IndividualUKPostcodeController @Inject()(
...
+        val json = Json.obj(
+          "form" -> preparedForm,
+          "mode" -> mode
+        )
 
-      renderer.render("individualUKPostcode.njk", json).map(Ok(_))
+        renderer.render("individualUKPostcode.njk", json).map(Ok(_))
+      } else {
+        Future.successful(Redirect(routes.WhatIsYourAddressController.onPageLoad(NormalMode)))
+      }
```
Comment (daniel-manning) [inline app/controllers/IndividualUKPostcodeController.scala:67]: we can push this into the routing instead

File: app/controllers/SelectAddressController.scala [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      } recover {
+        case _: Exception => Redirect(routes.WhatIsYourAddressUkController.onPageLoad(NormalMode))
+      }
+  }
+
+  private def formatAddress(address: AddressLookup) = {
+    val addressLine1 = address.addressLine1.fold("")(a => s"$a, ")
+    val addressLine2 = address.addressLine2.fold("")(a => s"$a, ")
+    val addressLine3 = address.addressLine3.fold("")(a => s"$a, ")
+    val addressLine4 = address.addressLine4.fold("")(a => s"$a, ")
```
Comment (daniel-manning) [inline app/controllers/SelectAddressController.scala:129]: `Seq(addressLine1, addressLine2, addressLine3, addressLine4).flatten.mkString(", ")`

File: app/models/AddressLookup.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            case 0 =>
+              (None, None, None, None)
+            case 1 =>
+              (Some(lines.head), None, None, None)
+            case 2 =>
+              (Some(lines.head), Some(lines(1)), None, None)
+            case 3 =>
+              (Some(lines.head), Some(lines(1)), Some(lines(2)), None)
+            case numberOfLines if numberOfLines >= 4 => (Some(lines.head), Some(lines(1)), Some(lines(2)), Some(lines(3)))
+          }
```
Comment (daniel-manning) [inline app/models/AddressLookup.scala:49]: ouch! Can't think of a better way though

File: app/connectors/AddressLookupConnector.scala (hunk 1 above)
Comment (helensoniega) [reply app/connectors/AddressLookupConnector.scala:47]: Changed it so we filter out addresses with undefined address lines 1 and 2.

File: test/navigation/NavigatorSpec.scala [hunk 7]
```diff
@@ -349,7 +352,11 @@ class NavigatorSpec extends SpecBase with ScalaCheckPropertyChecks with Generato
         }
       }
 
-      "must go from the Do You Live in the UK page for people who answer yes to What is your postcode?" in {
+      "must go from the Do You Live in the UK page for people who answer yes to What is your postcode? - " +
```
Comment (daniel-manning) [inline test/navigation/NavigatorSpec.scala:355]: somethings gone wrong in your quote " - {

File: test/navigation/NavigatorSpec.scala [hunk 8]
```diff
@@ -365,7 +372,27 @@ class NavigatorSpec extends SpecBase with ScalaCheckPropertyChecks with Generato
         }
       }
 
-      "must go from the Do You Live in the UK page for people who do to the What is your address page" in {
+      "must go from the Do You Live in the UK page for people who answer yes to What is your address? - " +
```
Comment (daniel-manning) [inline test/navigation/NavigatorSpec.scala:375]: here too

**Pull request merged**